### ISSUE TITLE
Fine-tune SAM on challenging cell segmentation images

### ISSUE THREAD

**Issue by sleeping4cat** [facebookresearch/segment-anything#456]
Recently, there was a slew of tutorials and research papers available for ```SAM``` with which you can fine-tune SAM for pretty much any task. But, I was recently trying to fine-tune the decider of SAM on my cell segmentation **(binary classification)** problem and it gave me an error. Most tutorials are made in mind for datasets where the segmentation part can be easily encapsulated in a bounding box but my image can't be done as that. As there are maybe 30 different cells in one image and it is quite difficult to place bounding box around those cells. 

The problem is not training SAM as SAM easily trains on my images rather it it is the moment of plotting the predicted masks by the model. I have tried experimenting with a few code and wrote some functions as well but they all fails. That's why, I am turning to the community if they could provide some support/guidance on this issue. It'll greatly help me. 

**References**
* https://medium.com/@rekalantar/segment-anything-model-sam-for-medical-image-segmentation-9344ba57f2ca
* https://github.com/NielsRogge/Transformers-Tutorials/blob/master/SAM/Run_inference_with_MedSAM_using_HuggingFace_Transformers.ipynb
* https://github.com/NielsRogge/Transformers-Tutorials/blob/master/SAM/Fine_tune_SAM_(segment_anything)_on_a_custom_dataset.ipynb

**Image in my Dataset (Mask)**
![11](https://github.com/facebookresearch/segment-anything/assets/112309211/987f10d3-b681-432c-b8a7-3ba6e651ef43)
**Image Copyright belongs to me and my team**

**Comment by lxr-1204** [comment]
Hello, I also encountered such a problem. I have two ideas: 
1. Use the image algorithm to generate the minimum circumscribed rectangle according to each mask, such as cv2.findContours; but using this method may cause problems in the output of the network. 
2. Convert the binary mask into logits, and use SAM as a mask, but I encountered obstacles when dealing with this step. For example the following picture
If you have similar problems, please discuss with me。

**Comment by mburaksayici** [comment]
What you can do is convert binary mask that contains multiple objects to polygons:
```

    def mask_to_polygons(self, mask, max_obj=10):
        contours, hierarchy = cv2.findContours(
            (mask * 255).astype(np.uint8), cv2.RETR_EXTERNAL, cv2.CHAIN_APPROX_SIMPLE
        )
        confidence = np.nanmean(np.where(mask != 0, mask, np.nan))

        polygons = []
        for cnt in contours:
            # Approximate the contour to a polygon
            epsilon = 0.001 * cv2.arcLength(cnt, True)
            approx = cv2.approxPolyDP(cnt, epsilon, True)

            # If polygon has at least 3 points, add it to list
            if len(approx) >= 3:
                masks = []
                for point in approx:
                    masks.append({"x": int(point[0][0]), "y": int(point[0][1])})
                polygon = {
                    "mask": masks,
                    "prediction_class": 1,
                    "confidence": float(confidence),
                }
                polygons.append(polygon)

        polygons = polygons[:max_obj]
        return polygons

```


and follow others.

**Comment by AymanTalkhani** [comment]
@sleeping4cat @lxr-1204 We've actually built out a platform that finetunes a research variant of SAM for labeling purposes at [InstaLabel](https://www.instalabel.ai/). All you need to do is upload your dataset, select and label a few samples you'd like to finetune the model on in our platform, and hit the 'Train annotator' button. You can then head over to the 'Annotate' tab, click 'Instant Annotation' and have the finetuned model label the rest of your dataset for you. You can even crowdsource the correction of the AI predictions for verified ground truth labels. Should be useful for a lot of custom computer vision labeling use cases. Hope this helps!

**Comment by MatiasMingo** [comment]
Hi!

Could any of you solve this issue? I'm facing the same problems with vegetation segmentation and classification.

Any information would be much appreciated
